translator/trace/zipkinv1_thrift_to_protospan.go

```diff
@@ -0,0 +1,211 @@
+		ipv6 = net.IP(e.Ipv6).String()
+	}
+	return &endpoint{
+		ServiceName: e.ServiceName,
+		IPv4:        ipv4,
+		IPv6:        ipv6,
+		Port:        int32(e.Port),
+	}
+}
+
+var trueByteSlice = []byte{1}
+
+func zipkinV1ThriftBinAnnotationsToOCAttributes(ztBinAnnotations []*zipkincore.BinaryAnnotation) (attributes *tracepb.Span_Attributes, localComponent string) {
+	if len(ztBinAnnotations) == 0 {
+		return nil, ""
+	}
+
+	attributeMap := make(map[string]*tracepb.AttributeValue)
+	for _, binaryAnnotation := range ztBinAnnotations {
+		pbAttrib := &tracepb.AttributeValue{}
+		binAnnotationType := binaryAnnotation.AnnotationType
+		switch binaryAnnotation.AnnotationType {
+		case zipkincore.AnnotationType_BOOL:
+			vBool := bytes.Equal(binaryAnnotation.Value, trueByteSlice)
+			pbAttrib.Value = &tracepb.AttributeValue_BoolValue{BoolValue: vBool}
+		case zipkincore.AnnotationType_BYTES:
+			bytesStr := base64.StdEncoding.EncodeToString(binaryAnnotation.Value)
+			pbAttrib.Value = &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{Value: bytesStr}}
+		case zipkincore.AnnotationType_DOUBLE:
+			var d float64
+			if err := bytesToNumber(binaryAnnotation.Value, &d); err != nil {
+				pbAttrib.Value = strAttributeForError(err)
+			} else {
+				pbAttrib.Value = &tracepb.AttributeValue_DoubleValue{DoubleValue: d}
+			}
+		case zipkincore.AnnotationType_I16:
+			var i int16
+			if err := bytesToNumber(binaryAnnotation.Value, &i); err != nil {
+				pbAttrib.Value = strAttributeForError(err)
+			} else {
+				pbAttrib.Value = &tracepb.AttributeValue_IntValue{IntValue: int64(i)}
+			}
+		case zipkincore.AnnotationType_I32:
+			var i int32
+			if err := bytesToNumber(binaryAnnotation.Value, &i); err != nil {
+				pbAttrib.Value = strAttributeForError(err)
+			} else {
+				pbAttrib.Value = &tracepb.AttributeValue_IntValue{IntValue: int64(i)}
+			}
+		case zipkincore.AnnotationType_I64:
+			var i int64
+			if err := bytesToNumber(binaryAnnotation.Value, &i); err != nil {
+				pbAttrib.Value = strAttributeForError(err)
+			} else {
+				pbAttrib.Value = &tracepb.AttributeValue_IntValue{IntValue: i}
+			}
+		case zipkincore.AnnotationType_STRING:
+			pbAttrib.Value = &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{Value: string(binaryAnnotation.Value)}}
+		default:
+			err := fmt.Errorf("unknown zipkin v1 binary annotation type (%d)", int(binAnnotationType))
+			pbAttrib.Value = strAttributeForError(err)
+		}
+
+		key := binaryAnnotation.Key
+		if key == zipkincore.LOCAL_COMPONENT {
+			// TODO: (@pjanotti) add reference to OpenTracing and change related tags to use them
+			key = "component"
+			localComponent = string(binaryAnnotation.Value)
+		}
+
+		attributeMap[key] = pbAttrib
+	}
+
+	attributes = &tracepb.Span_Attributes{
+		AttributeMap: attributeMap,
+	}
+	return attributes, localComponent
+}
+
+func bytesToNumber(b []byte, number interface{}) error {
+	buf := bytes.NewReader(b)
```

Comment:
This seems quite expensive for just the flexibility of writing to various types. I think we can make this simple by
```go
func toUint64WithPadding(b []byte) (uint64, error) {
        if len(b) == 0 {
            return 0, errNoBytes
        }
	bs := b[:]
	if diff := 8 - len(bs); diff > 0 {
		// Left pad the byte slice to maxSizeOf(uint64AsBytes) i.e. 0x7F FF FF FF FF FF FF FF.
		bs = append(bytes.Repeat([]byte{0x00}, diff), bs...)
	}
	return binary.BigEndian.Uint64(bs), nil
}
```

and then casting to the various values values e.g.
```go
      if i, err := toUint64WithPadding(binaryAnnotation.Value); err == nil {
           pAttrib.Value = &tracepb.AttributeValue_IntValue{IntValue: int64(i)}
      } else {
           pbAttrib.Value = strAttributeForError(err)
```